src/coreclr/jit/lower.cpp

```diff
@@ -2002,21 +2002,56 @@ void Lowering::LowerArgsForCall(GenTreeCall* call)
 }
 
 #if defined(TARGET_X86) && defined(FEATURE_IJW)
+//------------------------------------------------------------------------
+// LowerSpecialCopyArgs: Lower special copy arguments for P/Invoke IL stubs
+//
+// Arguments:
+//    call - the call node
+//
+// Notes:
+//    This method is used for P/Invoke IL stubs on x86 to handle arguments with special copy semantics.
+//    In particular, this method implements copy-constructor semantics for managed-to-unmanaged IL stubs
+//    for C++/CLI. In this case, the managed argument is passed by (managed or unmanaged) pointer in the
+//    P/Invoke signature with a speial modreq, but is passed to the unmanaged function by value.
+//    The value passed to the unmanaged function must be created through a copy-constructor call copying from
+//    the original source argument.
+//    We assume that the IL stub will be generated such that the following holds true:
+//       - If an argument to the IL stub has the special modreq, then its corresponding argument to the
+//         unmanaged function will be passed as the same argument index. Therefore, we can introduce the copy call
+//         from the original source argument to the argument slot in the unmanaged call.
 void Lowering::LowerSpecialCopyArgs(GenTreeCall* call)
 {
     // We only need to use the special copy helper on P/Invoke IL stubs
     // for the unmanaged call.
     if (comp->opts.jitFlags->IsSet(JitFlags::JIT_FLAG_IL_STUB) && comp->compMethodRequiresPInvokeFrame() &&
         call->IsUnmanaged())
     {
-        unsigned argIndex = 0;
+        // Unmanaged calling conventions on Windows x86 are passed in reverse order
+        // of managed args, so we need to count down the number of args.
+        // If the call is thiscall, we need to account for the this parameter,
+        // which will be first in the list.
+        // The this parameter is always passed in registers, so we can ignore it.
+        unsigned argIndex = call->gtArgs.CountUserArgs() - 1;
         for (CallArg& arg : call->gtArgs.Args())
         {
             if (!arg.IsUserArg())
             {
                 continue;
             }
 
+            if (call->GetUnmanagedCallConv() == CorInfoCallConvExtension::Thiscall &&
+                argIndex == call->gtArgs.CountUserArgs() - 1)
+            {
+                assert(arg.GetNode()->OperIs(GT_PUTARG_REG));
+                continue;
+            }
+
+            if (argIndex >= comp->info.compILargsCount)
+            {
+                argIndex--;
+                continue;
+            }
```

Comment:
I think you need some handling for retbufs as well.

Reply:
Ah, I guess not... Should be skipped by `IsUserArg`

Reply:
When is this `argIndex >= comp->info.compILargsCount` true?

Reply:
This fails on the LCID interop tests, in which we introduce an additional argument into the native call that isn't present in the managed P/Invoke signature.

The C++/CLI compiler will never emit a P/Invoke that uses both copy constructors and introduces the LCID argument.

Reply:
I guess one way to be completely explicit about it would be to check `m_specialCopyArgs != nullptr` above and add an `assert(call.CountUserArgs() == comp->info.compILargsCount)`, but this looks ok to me as well.

Reply:
I'll add the assert. Better to be explicit about it, and adding that check will probably help jit throughput (as we won't iterate through the args unless we actually are in a scenario where we may do something).

Reply:
I'll add the assert. Better to be explicit about it, and adding that check will probably help jit throughput (as we won't iterate through the args unless we actually are in a scenario where we may do something).